File: Java/프로그래머스/3/43163. 단어 변환/단어 변환.java
class Solution {
    boolean[] visited;
    int count;
    int min = Integer.MAX_VALUE;
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        visited = new boolean[words.length];
        boolean isSame = false;
        for (String word : words) {
            if (word.equals(target)) {
                isSame = true;
            }
        }
        if (!isSame) return 0;

        dfs(begin, target, words, count);
        if (min != Integer.MAX_VALUE) answer = min;
        return answer;
    }

    void dfs(String begin, String target, String[] words, int count) {
        if (begin.equals(target)) {
            min = Math.min(min, count);
            return;
        }

        for (int i = 0; i < words.length; i++) {
            if (visited[i]) continue;

            if (isPossible(begin, words[i])) {
                visited[i] = true;
                dfs(words[i], target, words, count + 1);
                visited[i] = false;
            }
        }
    }

    boolean isPossible(String begin, String target) {
        int count = 0;
        for (int i = 0; i < begin.length(); i++) {
            if (begin.charAt(i) != target.charAt(i)) count++;
        }
        return count == 1;
    }
}

File: Python/프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque


def solution(begin, target, words):
    answer = 0
    q = deque([(0, begin)])
    visited = [False] * len(words)

    while q:
        cnt, cur = q.popleft()

        if cur == target:
            return cnt

        for i in range(len(words)):
            if change(cur, words[i]) and not visited[i]:
                q.append((cnt + 1, words[i]))
                visited[i] = True
            else:
                continue

    return answer


def change(current, target):
    cnt = 0

    for i in range(len(current)):
        if current[i] == target[i]:
            cnt += 1

    if cnt == len(current) - 1:
        return 1
    else:
        return 0